### PR TITLE
Mapi 1140

### PR DESCRIPTION
**Pull request by alexduf**
- Move classes where they are more appropriate (````Tag````, ````Tags````, ````Push```` and ````PushResult````)
 - Create a new type ````Destination````, that is easier to manage than ````Either[Set[Topic], UserId]````
 - Modify the endpoint to ignore the potential topic passed as a parameter
 - Add a unit test for a collection of topics
 - Cleanup the test logs as the awsclient is way too verbose

### CODE REVIEW

File: notification/app/notification/models/Destination.scala [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package notification.models
+
+import models.{UserId, Topic}
+
+object Destination {
+  type Destination = Either[Set[Topic], UserId]
+
+  def apply(userId: UserId): Destination = Right(userId)
+  def apply(topic: Topic): Destination = Left(Set(topic))
+  def apply(topics: Set[Topic]): Destination = Left(topics)
```
Comment (agulowaty) [inline notification/app/notification/models/Destination.scala]: could we push to empty list of topics ?

Comment (alexduf) [inline notification/app/notification/models/Destination.scala]: That's a good point, I'll add a check in the controller

File: notification/conf/routes [hunk 2]
```diff
@@ -1,3 +1,4 @@
 GET         /healthcheck                        @notification.controllers.Main.healthCheck
 POST        /push/user/:userId                  @notification.controllers.Main.pushUser(userId: java.util.UUID)
+POST        /push/topic                         @notification.controllers.Main.pushTopics
```
Comment (agulowaty) [inline notification/conf/routes]: what should be in the body ?

File: notification/app/notification/models/Push.scala [hunk 3]
```diff
@@ -0,0 +1,9 @@
+package notification.models
+
+import models.Notification
+import notification.models.Destination.Destination
+import play.api.libs.json._
+
+case class Push(notification: Notification, destination: Destination) {
+  def tagQuery: Option[String] = None
```
Comment (agulowaty) [inline notification/app/notification/models/Push.scala]: why None ?

File: notification/app/notification/controllers/Main.scala [hunk 4]
```diff
@@ -55,8 +56,11 @@ final class Main @Inject()(
     }
   }
 
-  def pushTopic(topic: Topic): Action[Notification] = AuthenticatedAction.async(BodyJson[Notification]) { request =>
-    val push = Push(request.body, Left(topic))
+  @deprecated("A push notification can be sent to multiple topics, this is for backward compatibility only", since = "07/12/2015")
+  def pushTopic(topic: Topic): Action[Notification] = pushTopics
+
+  def pushTopics: Action[Notification] = AuthenticatedAction.async(BodyJson[Notification]) { request =>
+    val push = Push(request.body, Left(request.body.topic))
```
Comment (agulowaty) [inline notification/app/notification/controllers/Main.scala]: Field `.topic` should be pluralised in: https://github.com/guardian/mobile-n10n/blob/f2b89e81b5a1b32c08d4da604641410f0ef8f1bd/common/src/main/scala/models/Notification.scala#L15

File: notification/app/notification/services/WindowsNotificationSender.scala [hunk 5]
```diff
@@ -30,9 +33,21 @@ class WindowsNotificationSender(hubClient: NotificationHubClient, configuration:
...
 
-
-  private def getCounts(destination: Either[Topic, UserId]): Future[RepositoryResult[Int]] = destination match {
-    case Left(topic: Topic) => topicSubscriptionsRepository.count(topic)
+  private def count(destination: Destination): Future[RepositoryResult[Int]] = destination match {
+    case Left(topics: Set[Topic]) => sumOf(topics)
     case Right(_: UserId) => Future.successful(RepositoryResult(1))
   }
+
+  private def sumOf(topics: Set[Topic]): Future[RepositoryResult[Int]] = {
```
Comment (agulowaty) [inline notification/app/notification/services/WindowsNotificationSender.scala]: `sumOfSuccessfull` ?

File: notification/test/notification/services/AzureRawPushConverterSpec.scala [hunk 6]
```diff
@@ -31,17 +31,22 @@ class AzureRawPushConverterSpec extends Specification with Mockito {
...
+    "Convert a list of topics into a list of tag" in new PushConverterScope {
+      val topics = Set(Topic(Breaking, "uk"), Topic(Breaking, "us"))
+      val expected = Tags(Set(Tag("topic:Base16:627265616b696e672f756b"), Tag("topic:Base16:627265616b696e672f7573")))
+      azureRawPushConverter.toTags(Destination(topics)) shouldEqual Some(expected)
     }
   }
 
   trait PushConverterScope extends Scope {
-    def configuration = {
+    def configuration: Configuration = {
```
Comment (agulowaty) [inline notification/test/notification/services/AzureRawPushConverterSpec.scala]: could be val

File: notification/app/notification/controllers/Main.scala (hunk 4 above)
Comment (alexduf) [inline notification/app/notification/controllers/Main.scala]: I know but the client is already released.

File: notification/app/notification/services/WindowsNotificationSender.scala (hunk 5 above)
Comment (alexduf) [inline notification/app/notification/services/WindowsNotificationSender.scala]: It's either the sum of everything or it returns an error, so sumOf seems relevant.

File: notification/app/notification/models/Destination.scala (hunk 1 above)
Comment (alexduf) [inline notification/app/notification/models/Destination.scala]: the check has been added

Comment (agulowaty) [inline notification/app/notification/models/Destination.scala]: if at least one topic is required to define Destination, maybe vararg like
```
def apply(topic: Topic, topics: Topic*)
```
will suite better ?

**Comment by agulowaty** [comment]
:+1:

**Pull request merged**